### DISCUSSION TITLE
How to print all Loihi simulator compiler output over the entire simulation graph.

### DISCUSSION THREAD

**Discussion by russelljjarvis** [lava-nc/lava#240] [Q&A]
I have made an elaborate cortical simulation and I am trying to print out the Loihi simulation code.

I have found how to print out parts of the graph using object attributes (see code example below), but I wonder how I can print out all relevant compiler information all at once not just for the `cortical_column_monitor` but for every network monitor?

``` python
c = Compiler()
exe = c.compile(cortical_column_monitor, Loihi1SimCfg())
print(next(iter(exe.py_builders)).proc_params,
                 other_column.proc_params)

print(next(iter(exe.py_builders)).weights)
print(next(iter(exe.py_builders)).in_ports)
print(next(iter(exe.py_builders)).weight_exp)
print(next(iter(exe.py_builders)).a_out)
print(next(iter(exe.py_builders)).num_weight_bits)
```

@samalika
@wangrunchun

**Comment by awintel** [comment]
You are deep into the domain of interacting with "private" data structures at the user level.

What objective are you pursuing in printing out 'builder' fields? 

What do you define as 'simulation code' and 'simulation graph'? 
What is 'relevant' compiler information (relevant for what)?
Are you trying to gain access to timeseries data of state variables during simulation? 
Are you trying to visualize your static network structure?

**Reply by russelljjarvis** [reply]
Hi @awintel, 

Sorry I dropped the ball on this question. 

> You are deep into the domain of interacting with "private" data structures at the user level.
> 
> What objective are you pursuing in printing out 'builder' fields?

--I am thinking about how I could map an different type of FPGA hardware backend onto Lava. 
 
> What do you define as 'simulation code' and 'simulation graph'? What is 'relevant' compiler information (relevant for what)? Are you trying to gain access to timeseries data of state variables during simulation? Are you trying to visualize your static network structure?

I had just picked up the concept that Lava code can somehow be represented as a graph of interacting components and I don't have a clear idea about how the simulation graph works yet. I have also noticed that to run a simulation on a whole network, you only need to call `run` or `simulate` on just an element of the whole network. 

Are you trying to visualise your static network structure? -- No, I wasn't but that is independently a very important goal of mine. I am working on a streamlit [front end](https://share.streamlit.io/russelljjarvis/potjansvis/main/app.py) for Lava which is different to my goal of exploring if an FPGA backend is possible. Please elaborate on about how to visualize the static network structure if you know of a way.

Are you trying to gain access to timeseries data of state variables during simulation? -- No.

I am really just interested in what is Lava final representation of a network simulation to be run, before that simulation is consumed by any of the available hardware backends.

**Comment by samalika** [comment]
Hey, I can see, I think we can do something similar using networkX weighted graphs which has added al attributes
________________________________
From: Russell Jarvis ***@***.***>
Sent: 02 May 2022 16:29
To: lava-nc/lava ***@***.***>
Cc: Samalika Perera ***@***.***>; Mention ***@***.***>
Subject: [lava-nc/lava] How to print all Loihi simulator compiler output over the entire simulation graph. (Discussion #240)


I have made an elaborate cortical simulation and I am trying to print out the Loihi simulation code.

I have found how to print out parts of the graph using object attributes (see code example below), but I wonder how I can print out all relevant compiler information all at once not just for the cortical_column_monitor but for every network monitor?

c = Compiler()
exe = c.compile(cortical_column_monitor, Loihi1SimCfg())
print(next(iter(exe.py_builders)).proc_params,
                 other_column.proc_params)

print(next(iter(exe.py_builders)).weights)
print(next(iter(exe.py_builders)).in_ports)
print(next(iter(exe.py_builders)).weight_exp)
print(next(iter(exe.py_builders)).a_out)
print(next(iter(exe.py_builders)).num_weight_bits)

@samalika<https://github.com/samalika>
